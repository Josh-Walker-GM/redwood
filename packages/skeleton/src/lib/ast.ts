import fs from 'fs'

import { parse } from '@babel/parser'
import traverse from '@babel/traverse'
import {
  assertJSXOpeningElement,
  assertJSXIdentifier,
  assertIdentifier,
  isJSXAttribute,
  isJSXSpreadAttribute,
  JSXElement,
  ExportNamedDeclaration,
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
  File,
  VariableDeclarator,
  ExportDefaultDeclaration,
} from '@babel/types'

export function getASTFromCode(code: string) {
  return parse(code, {
    sourceType: 'unambiguous',
    // TODO: Check these plugin options are optimal, I doubt they are
    plugins: [
      'jsx',
      'typescript',
      'nullishCoalescingOperator',
      'objectRestSpread',
    ],
  })
}

export function getASTFromFile(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return getASTFromCode(code)
}

export function getSpecificNamedExportDeclarations(ast: File, names: string[]) {
  const desiredNamedExports: Map<string, VariableDeclarator> = new Map()
  const allNamedExports: ExportNamedDeclaration[] = []
  traverse(ast, {
    ExportNamedDeclaration: (path) => {
      allNamedExports.push(path.node)
    },
  })
  allNamedExports.forEach((namedExport) => {
    if (
      namedExport.declaration != null &&
      isVariableDeclaration(namedExport.declaration) &&
      isVariableDeclarator(namedExport.declaration.declarations[0]) &&
      isIdentifier(namedExport.declaration.declarations[0].id) &&
      names.includes(namedExport.declaration.declarations[0].id.name)
    ) {
      desiredNamedExports.set(
        namedExport.declaration.declarations[0].id.name,
        namedExport.declaration.declarations[0]
      )
    }
  })
  return desiredNamedExports
}

export function getExportDefaultDeclaration(ast: File) {
  let defaultExport: ExportDefaultDeclaration | undefined
  traverse(ast, {
    ExportDefaultDeclaration: (path) => {
      defaultExport = path.node
    },
  })
  return defaultExport
}

export function getVariableDeclarator(ast: File, name: string) {
  let variableDeclarator: VariableDeclarator | undefined
  traverse(ast, {
    VariableDeclarator: (path) => {
      if (isIdentifier(path.node.id) && path.node.id.name === name) {
        variableDeclarator = path.node
      }
    },
  })
  return variableDeclarator
}

export function getJSXElementAttributes(
  element: JSXElement
): Map<string, string> {
  const attributes: Map<string, string> = new Map()
  assertJSXOpeningElement(element.openingElement)
  if (element.openingElement.attributes.length > 0) {
    element.openingElement.attributes.forEach((attribute) => {
      if (isJSXAttribute(attribute)) {
        assertJSXIdentifier(attribute.name)
        const key = attribute.name.name
        let value: string
        switch (attribute.value?.type) {
          case 'StringLiteral':
            value = attribute.value.value
            break
          case 'JSXExpressionContainer':
            assertIdentifier(attribute.value.expression)
            value = attribute.value.expression.name
            break
          case undefined: // Accept this because it could be a attribute with no value, i.e. notfound
            value = 'true'
            break
          default:
            throw new Error(
              `Unknown value type found while trying to process the ${key} property when extracting all attributes`
            )
        }
        attributes.set(key, value)
      } else if (isJSXSpreadAttribute(attribute)) {
        throw new Error(
          'JSXSpreadAttribute type is not yet supported when extracting all attributes'
        )
      } else {
        throw new Error(
          'Uknown attribute type found while trying to extract all attributes'
        )
      }
    })
  }
  return attributes
}
