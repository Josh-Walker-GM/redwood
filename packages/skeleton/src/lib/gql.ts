import { parse } from 'graphql'

export function parseGraphQL(gql: string) {
  return parse(gql)
}

export function getGraphQLQueryName(gql: string): string | undefined {
  const ast = parse(gql)
  for (const def of ast.definitions) {
    if (def.kind === 'OperationDefinition' && def.operation === 'query') {
      return def.name?.value
    }
  }
  return undefined
}
