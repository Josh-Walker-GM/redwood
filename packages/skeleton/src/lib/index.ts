import fs from 'fs'

import { parse } from '@babel/parser'

export function getFileAST(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return parse(code, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
  })
}
