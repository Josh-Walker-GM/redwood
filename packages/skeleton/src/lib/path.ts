import path from 'path'

import { getBaseDir, getConfigPath } from '@redwoodjs/internal/dist/index'

/**
 * Determines the root path of a redwood project which contains the provided path
 * @param pathWithinProject A path within a redwood project
 * @returns The root path of the associated project
 */
export function getRootPath(pathWithinProject: string) {
  return pathWithinProject
    ? path.dirname(getConfigPath(pathWithinProject))
    : getBaseDir()
}
