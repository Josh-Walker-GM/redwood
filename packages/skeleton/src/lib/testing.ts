import { ensurePosixPath } from '@redwoodjs/internal/dist/paths'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
