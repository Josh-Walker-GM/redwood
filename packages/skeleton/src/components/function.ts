import fs from 'fs'
import path from 'path'

import traverse from '@babel/traverse'
import type { ExportNamedDeclaration } from '@babel/types'
import {
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
} from '@babel/types'

import { getPaths } from '@redwoodjs/internal/dist/paths'

import { getASTFromFile } from '../lib/ast'

import { RedwoodSkeleton } from './base'
import type { RedwoodProject } from './project'

export class RedwoodFunction extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  constructor(filepath: string) {
    super(filepath)

    const ast = getASTFromFile(this.filepath)
    const namedExports: ExportNamedDeclaration[] = []
    traverse(ast, {
      ExportNamedDeclaration: (path) => {
        namedExports.push(path.node)
      },
    })
    const handlerExport = namedExports.find((node) => {
      return (
        isVariableDeclaration(node.declaration) &&
        isVariableDeclarator(node.declaration.declarations[0]) &&
        isIdentifier(node.declaration.declarations[0].id) &&
        node.declaration.declarations[0].id.name === 'handler'
      )
    })
    if (handlerExport === undefined) {
      this.errors.push('No "handler" export found')
    }
  }

  getInformation(): string {
    return '' // I don't think there is any useful information about the function, so return an empty string
  }
}

export function extractFunction(filepath: string): RedwoodFunction {
  return new RedwoodFunction(filepath)
}

export function extractFunctions(
  project: RedwoodProject | undefined = undefined
): RedwoodFunction[] {
  const functions: RedwoodFunction[] = []

  const functionsPath = project
    ? getPaths(project.filepath).api.functions
    : getPaths().api.functions

  if (!fs.existsSync(functionsPath)) {
    return functions
  }

  // TODO: Confirm this the condition to detect a function
  // Functions must be defined within the functions directory (at the top level?)
  const getFunctionFiles = (directory: string) => {
    const functionFiles: string[] = []
    const directoryContents = fs.readdirSync(directory)
    directoryContents.forEach((content) => {
      const stat = fs.lstatSync(path.join(directory, content))
      if (stat.isFile()) {
        functionFiles.push(path.join(directory, content))
      }
    })
    return functionFiles
  }

  const layoutFiles = getFunctionFiles(functionsPath)
  layoutFiles.forEach((layoutPath) => {
    functions.push(extractFunction(layoutPath))
  })

  return functions
}
