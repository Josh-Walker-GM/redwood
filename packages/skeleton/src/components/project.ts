import fs from 'fs'
import path from 'path'

import chalk from 'chalk'

import { getPaths } from '@redwoodjs/internal/dist/index'

import { getRootPath } from '../lib/path'

import { RedwoodSkeleton } from './base'
import { extractCells, RedwoodCell } from './cell'
import { extractFunctions, RedwoodFunction } from './function'
import { extractLayouts, RedwoodLayout } from './layout'
import { extractPages, RedwoodPage } from './page'
import { extractRouters, RedwoodRouter } from './router'
import { extractSides, RedwoodSide } from './side'

/**
 * Used to enumerate either JS or TS project types
 */
export enum RedwoodProjectType {
  TYPESCRIPT = 'typescript',
  JAVASCRIPT = 'javascript',
}

export class RedwoodProject extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly type: RedwoodProjectType

  private sides?: RedwoodSide[] | undefined
  private cells?: RedwoodCell[] | undefined
  private routers?: RedwoodRouter[] | undefined
  private layouts?: RedwoodLayout[] | undefined
  private pages?: RedwoodPage[] | undefined
  private functions?: RedwoodFunction[] | undefined

  public static getProject({
    pathWithinProject = '',
    full = false,
    readFromCache = true,
    insertIntoCache = true,
  }: {
    pathWithinProject?: string
    full?: boolean
    readFromCache?: boolean
    insertIntoCache?: boolean
  } = {}) {
    const projectCache = RedwoodProjectsCache.getInstance().projects
    const rootPath = getRootPath(pathWithinProject)

    if (readFromCache) {
      const cachedProject = projectCache.get(rootPath)
      if (cachedProject) {
        return cachedProject
      }
    }

    const project = new RedwoodProject(rootPath, full)
    if (insertIntoCache) {
      projectCache.set(project.filepath, project)
    }
    return project
  }

  private constructor(rootPath: string, full = false) {
    super(rootPath)

    // A project is typescript if we detect a tsconfig.json
    this.type =
      fs.existsSync(
        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
      ) ||
      fs.existsSync(
        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
      )
        ? RedwoodProjectType.TYPESCRIPT
        : RedwoodProjectType.JAVASCRIPT

    if (full) {
      this.sides = extractSides(this)
      this.cells = extractCells(this)
      this.layouts = extractLayouts(this)
      this.pages = extractPages(this)
      this.functions = extractFunctions(this)
      this.routers = extractRouters(this)
    }
  }

  getSides(forceExtract = false): RedwoodSide[] {
    if (forceExtract) {
      return extractSides(this)
    }
    return this.sides === undefined ? extractSides(this) : this.sides
  }

  getCells(forceExtract = false): RedwoodCell[] {
    if (forceExtract) {
      return extractCells(this)
    }
    return this.cells === undefined ? extractCells(this) : this.cells
  }

  getRouters(forceExtract = false): RedwoodRouter[] {
    if (forceExtract) {
      return extractRouters(this)
    }
    return this.routers === undefined ? extractRouters(this) : this.routers
  }

  getLayouts(forceExtract = false): RedwoodLayout[] {
    if (forceExtract) {
      return extractLayouts(this)
    }
    return this.layouts === undefined ? extractLayouts(this) : this.layouts
  }

  getPages(forceExtract = false): RedwoodPage[] {
    if (forceExtract) {
      return extractPages(this)
    }
    return this.pages === undefined ? extractPages(this) : this.pages
  }

  getFunctions(forceExtract = false): RedwoodFunction[] {
    if (forceExtract) {
      return extractFunctions(this)
    }
    return this.functions === undefined
      ? extractFunctions(this)
      : this.functions
  }

  // Diagnostics

  getInformation(): string {
    return '' // TODO: Implement, gets the info of all other components too
  }

  printInformation(): void {
    console.log(
      `${chalk.bgCyan('[Info]')}\t${this.name}\t${chalk.dim(this.filepath)}`
    )
  }

  printWarnings(cascade = false): void {
    if (this.warnings.length > 0) {
      const titleLine = `${chalk.bgYellow('[Warn]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const warningLines = this.warnings.map((warning, index) => {
        return ` (${index + 1}) ${warning}\n`
      })
      console.log(titleLine.concat('\n', ...warningLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printWarnings()
      })
      this.functions?.forEach((func) => {
        func.printWarnings()
      })
      this.layouts?.forEach((layout) => {
        layout.printWarnings()
      })
      this.pages?.forEach((page) => {
        page.printWarnings()
      })
      this.routers?.forEach((router) => {
        router.printWarnings()
        router.routes.forEach((route) => {
          route.printWarnings()
        })
      })
      this.sides?.forEach((side) => {
        side.printWarnings()
      })
    }
  }

  printErrors(cascade = false): void {
    if (this.errors.length > 0) {
      const titleLine = `${chalk.bgRed('[Error]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const errorLines = this.errors.map((error, index) => {
        return ` (${index + 1}) ${error}\n`
      })
      console.log(titleLine.concat('\n', ...errorLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printErrors()
      })
      this.functions?.forEach((func) => {
        func.printErrors()
      })
      this.layouts?.forEach((layout) => {
        layout.printErrors()
      })
      this.pages?.forEach((page) => {
        page.printErrors()
      })
      this.routers?.forEach((router) => {
        router.printErrors()
        router.routes.forEach((route) => {
          route.printErrors()
        })
      })
      this.sides?.forEach((side) => {
        side.printErrors()
      })
    }
  }
}

/**
 * Singelton class which contains a cache of {@link RedwoodProject} instances.
 *
 * This class is not exported and should only be available from static methods within {@link RedwoodProject}, in order to keep the surface of the skeleton API cleaner and leaner.
 */
class RedwoodProjectsCache {
  private static instance: RedwoodProjectsCache

  public projects: Map<string, RedwoodProject>

  private constructor() {
    this.projects = new Map()
  }

  public static getInstance(): RedwoodProjectsCache {
    if (!RedwoodProjectsCache.instance) {
      RedwoodProjectsCache.instance = new RedwoodProjectsCache()
    }
    return RedwoodProjectsCache.instance
  }
}
