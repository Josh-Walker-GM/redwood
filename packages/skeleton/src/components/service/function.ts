import path from 'path'

import {
  ExportNamedDeclaration,
  isArrowFunctionExpression,
  isFunctionExpression,
  isIdentifier,
  isObjectPattern,
  isObjectProperty,
  isVariableDeclaration,
  isVariableDeclarator,
} from '@babel/types'

import { RedwoodError, RedwoodErrorCode, RedwoodWarning } from '../diagnostic'
import { RedwoodProject } from '../project'
import type { RedwoodSDLMutation } from '../sdl/mutations'
import type { RedwoodSDLQuery } from '../sdl/query'
import { RedwoodSkeleton } from '../skeleton'

export class RedwoodServiceFunction extends RedwoodSkeleton {
  warnings: RedwoodWarning[] = []
  errors: RedwoodError[] = []

  readonly serviceName: string
  readonly parameters: string[] = []

  constructor(
    filepath: string,
    exportNamedDeclaration: ExportNamedDeclaration
  ) {
    // Extract the function name
    let name: string | undefined = undefined
    if (
      isVariableDeclaration(exportNamedDeclaration.declaration) &&
      exportNamedDeclaration.declaration.declarations[0] != null &&
      isVariableDeclarator(
        exportNamedDeclaration.declaration.declarations[0]
      ) &&
      isIdentifier(exportNamedDeclaration.declaration.declarations[0].id)
    ) {
      name = exportNamedDeclaration.declaration.declarations[0].id.name
    }

    super(filepath, name)
    this.serviceName = path.parse(this.filepath).name

    if (name === undefined) {
      this.errors.push({
        code: RedwoodErrorCode.GENERIC_PARSER_ERROR_JSTS,
        message: 'Could not determine the name of a named export',
      })
    }

    // Extract the function parameters
    if (
      isVariableDeclaration(exportNamedDeclaration.declaration) &&
      exportNamedDeclaration.declaration.declarations[0] != null &&
      isVariableDeclarator(
        exportNamedDeclaration.declaration.declarations[0]
      ) &&
      (isArrowFunctionExpression(
        exportNamedDeclaration.declaration.declarations[0].init
      ) ||
        isFunctionExpression(
          exportNamedDeclaration.declaration.declarations[0].init
        ))
    ) {
      const parameters =
        exportNamedDeclaration.declaration.declarations[0].init.params[0]
      if (isObjectPattern(parameters)) {
        parameters.properties.forEach((property) => {
          if (isObjectProperty(property) && isIdentifier(property.key)) {
            this.parameters.push(property.key.name)
          }
        })
      }
    }

    // Checks

    const associatedSDLOperation = this.getAssociatedSDLOperation()
    if (associatedSDLOperation) {
      let parametersMatch =
        associatedSDLOperation.parameters.length === this.parameters.length
      for (let i = 0; i < this.parameters.length; i++) {
        parametersMatch &&=
          this.parameters[i] === associatedSDLOperation.parameters[i]
      }
      if (!parametersMatch) {
        this.errors.push({
          code: RedwoodErrorCode.SERVICE_FUNCTION_PARAMETERS_DO_NOT_MATCH_SDL_OPERATION,
          message: `The parameters of the "${this.name}" function do not match those of the correspondind SDL operation "${associatedSDLOperation.name}"`,
        })
      }
    }
  }

  getAssociatedSDLOperation():
    | undefined
    | RedwoodSDLQuery
    | RedwoodSDLMutation {
    const sdl = RedwoodProject.getProject({ pathWithinProject: this.filepath })
      .getSDLs()
      .find((sdl) => {
        return sdl.name === this.serviceName
      })
    if (sdl) {
      const query = sdl.queries?.find((query) => {
        return query.name === this.name
      })
      const mutation = sdl.mutations?.find((mutation) => {
        return mutation.name === this.name
      })
      return query || mutation
    }
    return undefined
  }
}
