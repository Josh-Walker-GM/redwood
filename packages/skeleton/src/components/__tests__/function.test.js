import path from 'path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { extractFunctions } from '../function'
import { RedwoodProject } from '../project'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)

describe.each([
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
])('From within the %s fixture', (PROJECT_NAME) => {
  const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = PROJECT_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  it('returns the correct functions without a project', () => {
    const functions = extractFunctions(undefined).sort((a, b) =>
      a.filepath > b.filepath ? 1 : b.filepath > a.filepath ? -1 : 0
    )
    functions.forEach((aFunction) => {
      aFunction.filepath = stripAndFormatPathForTesting(
        aFunction.filepath,
        PROJECT_PATH
      )
    })
    expect(functions).toMatchSnapshot()
  })

  it('returns the correct functions with a project', () => {
    const project = RedwoodProject.getProject({
      pathWithinProject: PROJECT_PATH,
      readFromCache: false,
      insertIntoCache: false,
    })
    const functions = extractFunctions(project).sort((a, b) =>
      a.filepath > b.filepath ? 1 : b.filepath > a.filepath ? -1 : 0
    )
    functions.forEach((aFunction) => {
      aFunction.filepath = stripAndFormatPathForTesting(
        aFunction.filepath,
        PROJECT_PATH
      )
    })
    expect(functions).toMatchSnapshot()
  })
})
