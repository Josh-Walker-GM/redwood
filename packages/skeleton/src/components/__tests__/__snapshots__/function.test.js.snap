// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture returns the correct functions with a project: /api/src/functions/graphql.ts 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.ts",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the empty-project fixture returns the correct functions with a project: function count 1`] = `1`;

exports[`From within the empty-project fixture returns the correct functions without a project: /api/src/functions/graphql.ts 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.ts",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the empty-project fixture returns the correct functions without a project: function count 1`] = `1`;

exports[`From within the example-todo-main fixture returns the correct functions with a project: /api/src/functions/graphql.js 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.js",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct functions with a project: function count 1`] = `1`;

exports[`From within the example-todo-main fixture returns the correct functions without a project: /api/src/functions/graphql.js 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.js",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct functions without a project: function count 1`] = `1`;

exports[`From within the example-todo-main-with-errors fixture returns the correct functions with a project: /api/src/functions/graphql.js 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.js",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct functions with a project: function count 1`] = `1`;

exports[`From within the example-todo-main-with-errors fixture returns the correct functions without a project: /api/src/functions/graphql.js 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.js",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the example-todo-main-with-errors fixture returns the correct functions without a project: function count 1`] = `1`;

exports[`From within the test-project fixture returns the correct functions with a project: /api/src/functions/auth.ts 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/auth.ts",
  "name": "auth",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct functions with a project: /api/src/functions/graphql.ts 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.ts",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct functions with a project: function count 1`] = `2`;

exports[`From within the test-project fixture returns the correct functions without a project: /api/src/functions/auth.ts 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/auth.ts",
  "name": "auth",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct functions without a project: /api/src/functions/graphql.ts 1`] = `
RedwoodFunction {
  "errors": [],
  "filepath": "/api/src/functions/graphql.ts",
  "name": "graphql",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct functions without a project: function count 1`] = `2`;
